x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authz/AuthorizationService.java

```diff
@@ -173,69 +175,123 @@ private AuthorizationInfo getAuthorizationInfoFromContext() {
+            }
+        }
+    }
+
+    @Nullable
+    private static AuthorizationContext extractAuthorizationContext(ThreadContext threadContext, String childAction) {
+        final String parentAction = threadContext.getTransient(ORIGINATING_ACTION_KEY);
+        if (Strings.isNullOrEmpty(parentAction)) {
+            // No parent action
+            return null;
+        }
+        AuthorizationInfo authorizationInfo = threadContext.getTransient(AUTHORIZATION_INFO_KEY);
+        if (authorizationInfo == null) {
+            throw internalError(
+                "Attempt to authorize child action [" + childAction + "] of [" + parentAction + "] without existing authorization info"
+            );
+        }
+        if (isIndexAction(parentAction)) {
+            final IndicesAccessControl parentAccessControl = threadContext.getTransient(AuthorizationServiceField.INDICES_PERMISSIONS_KEY);
+            return new AuthorizationContext(parentAction, authorizationInfo, new IndexAuthorizationResult(true, parentAccessControl));
+        } else {
+            // We assume the parent is true here, but maybe we actually need to store the AuthorizationResult in a transient
+            return new AuthorizationContext(parentAction, authorizationInfo, new AuthorizationResult(true));
+        }
+    }
+
+    private String requireAuditId(Authentication authentication, String action, TransportRequest originalRequest) {
+        String auditId = AuditUtil.extractRequestId(threadContext);
+        if (auditId == null) {
+            // We would like to assert that there is an existing request-id, but if this is a system action, then that might not be
+            // true because the request-id is generated during authentication
+            if (isInternal(authentication.getUser())) {
+                auditId = AuditUtil.getOrGenerateRequestId(threadContext);
+            } else {
+                auditTrailService.get().tamperedRequest(null, authentication, action, originalRequest);
+                throw internalError(
+                    "Attempt to authorize action ["
+                        + action
+                        + "] for ["
+                        + authentication.getUser().principal()
+                        + "] without an existing request-id"
+                );
             }
         }
+        return auditId;
+    }
+
+    private static ElasticsearchSecurityException internalError(String message) {
+        // When running with assertions enabled (testing) kill the node so that there is a hard failure in CI
+        assert false : message;
+        // Otherwise (production) just throw an exception so that we don't authorize something incorrectly
+        return new ElasticsearchSecurityException(message);
+    }
+
+    private boolean checkOperatorPrivileges(
+        Authentication authentication,
+        String action,
+        TransportRequest originalRequest,
+        ActionListener<Void> listener
```

Comment:
nit: returning a value and also using a listener is confusing. Maybe take a `Consumer<Exception>` as the last parameter and return `void`.

Reply:
I don't think it should take a `Consumer` because then the consumer would need to set a local variable to know whether the authorize method should return.
We can throw an exception though,